File: cplusplus/medium/DecodeString.cpp
class Solution {
public:
    string decodeString(string s) {
    
        stack<char> st;
        
        for(auto &ch: s){
            if(ch == ']'){
                char t = st.top();
                st.pop();
                string repeated = "";
                while(t != '['){
                    repeated = t + repeated;
                    t = st.top();
                    st.pop();
                }


                t = st.top();
                st.pop();
                string time_str = "";
                while((t>='0') && (t<='9')){

                    time_str = t + time_str;
                    if(!st.empty()){
                        t = st.top();
                        st.pop();
                    }else{
                        break;
                    }

                }

                int times = stoi(time_str);
                if((t<'0') || (t>'9')){
                    st.push(t);
                }

                for(int i=0; i<times; ++i){
                    for(auto &tmp_r: repeated){
                        st.push(tmp_r);
                    }
                }
            }else{
                st.push(ch);
            }
        }
        
        string res = "";
        while(!st.empty()){
            char ch = st.top();
            res = ch + res; 
            st.pop();
        }
        return res;
        
    }
};

File: python/medium/DecodeString.py
class Solution(object):
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        if  not "[" in s:
            return  s
        
        times = 0
        start = 0
        queue = []
        ret = ""
        for idx, char in enumerate(s):
            if ord(char)>=ord("0") and ord(char) <= ord("9"):
                if not queue:
                    times = times*10 + int(char)
            elif char == "[":
                if not queue:
                    start = idx+1    
                queue.append("[")
            elif char == "]":
                queue.pop()
                if not queue:
                    tmp = self.decodeString(s[start:idx])
                    print(tmp)
                    for i in range(times):
                        ret += tmp
                    times = 0
            else:
                if not queue:
                    ret += char
        return ret